django/db/models/query.py

```diff
@@ -798,6 +798,23 @@ def exists(self):
             return self.query.has_results(using=self.db)
         return bool(self._result_cache)
 
+    def contains(self, obj):
+        """
+        Check if object exists in queryset.
+
+        If the QuerySet is already fully cached, check if object is in cache.
+        """
+        try:
+            if obj._meta.concrete_model != self.model._meta.concrete_model:
+                return False
+        except AttributeError:
+            raise NotImplementedError(
+                'QuerySet.contains only supports Model objects. You passed in a {}.'.format(type(obj))
+            )
+        if self._result_cache is None:
+            return self.filter(pk=obj.pk).exists()
+        return obj in self._result_cache
```

Comment:
This a O(n) scan in `self._result_cache`, not sure this is desirable.

Reply:
I did extensive testing on this, but only using localhost database. The result was clear that unless you have a VERY big data set, it will be faster scanning prefetched data than hitting the database one more time. If you are working on a very bit dataset and already did a prefetch, the extra overhead will likely be marginal to doing the actual data fetch and object mapping. Any developer who want's to optimize can still easily decide method just like in the past.

I feel confident that this behaviour will choose the fastest method for most cases. Still, for optimization a developer might want to be explicit about method, instead of calling .contains(obj).

Reply:
Another thing about `self._result_cache` which is problematic is the chaining of `values` (and `values_list`, ..) which will store non-model instances in there. That will result in this strange behavior

```python
entry = Entry.objects.get(pk=123)
values_queryset = Entry.objects.values('id')
values_queryset.contains(entry)  # True
list(values_queryset)
values_queryset.contains(entry)  # False
```

I would expect `.contains` to immediately error out `if not issubclass(self._iterable_class, ModelIterable)` or to support `values` and friends in a consistent way.

e.g.

```python
entry = Entry.objects.get(pk=123)
values_queryset = Entry.objects.filter(pk=123).values('id')
values_queryset.contains(entry)  # Crash
values_queryset.contains({'id': 123})  # True
values_queryset.contains({'id': 456})  # False
list(values_queryset)
values_queryset.contains({'id': 123})  # True
values_queryset.contains({'id': 456})  # False
```

If we want to go this way the best approach is likely to defer the validation of `obj` and the lookup creation to a new `BaseIterable.contains_lookup(obj)` class method so the implementation can be along the lines of

```python
class BaseIterable:
    ...
    @classmethod
    def contains_lookup(cls, queryset, obj):
        raise NotImplementedError

class ModelIterable(BaseIterable):
    ...
    @classmethod
    def contains_lookup(cls, queryset, obj):
        try:
            if obj._meta.concrete_model != queryset.model._meta.concrete_model:
                return False
        except AttributeError:
            raise ValueError(
                'QuerySet.contains only supports Model objects. You passed in a {}.'.format(type(obj))
            )
        return {'pk': obj.pk}

class ValuesIterable(BaseIterable):
    ...
    @classmethod
    def contains_lookup(cls, queryset, obj):
        if not isinstance(obj, dict):
            raise ValueError(...)
        names = {
            *query.extra_select,
            *query.values_select,
            *query.annotation_select,
        }
        if not set(obj) == names:
            raise ValueError(...)
        return obj

class QuerySet:
    ...
    def contains(self, obj):
        lookup = self._iterable_class.contains_lookup(self, obj)
        if self._result_cache is None:
            return self.filter(**lookup).exists()
        return obj in self._result_cache
```

Reply:
I'll be honest and admit I thought values() and values_list() didn't return actual querysets. Thinking about it, It's kind of obvious, because I've used them combined with .annotate() to group stuff.

Maybe it's better if someone else takes this patch? I'm clearly in over my head here. I understand your code above, but I can't really wrap my head around the implications at this time. I would probably go with ModelIterable only, and otherwise hit the database with an exists() query, because it feels safe.

Reply:
Erroring out immediately `if not issubclass(self._iterable_class, ModelIterable)` is likely the easiest way to go here. Just wanted to point out that we need to handle this case one way or another.

Reply:
Feels like it would be a weird behaviour to document and kind of a foot gun interface. Why would you use `values().contain` in the first place if its only a proxy to `.exists()`?

A clear error would guide a user on relying directly on `exists` instead. That's just my perception on the problem but maybe other folks have a different stance on the problem.

Reply:
I pushed a new version. I do agree that it's a weird scenario for calling `QuerySet.contains()`. However, there might be just cause. Say you have a grouped dataset using `qs.values('pk').annotate(sum=Sum('value'))`, and you want to check if a specific object is in there. If someone is used to `.contains()`, why force the use of `.filter(pk=obj.pk).exists()` if they don't care and it doesn't matter to them.

Yes, people can use this in ways that hits the database more than they expect, but that's always the case when people don't understand enough about the framework. I'll be the first to admit that I've not understood stuff like that though years of using Django.

Reply:
@charettes I should also note that if you check the discussion in the mailing list (referred in description above), you'll get the motivation for adding `.contains()`, even though it's sort of a proxy for `.exists()`. I say _sort of_, because it does more, and with a more comprehensive syntax.

Reply:
@charettes Do you still have objections here?

Reply:
yes I think the `values.contains` point should be discussed on the mailing list if we don't want to explicitly error out in this case. I would personally expect `values.contains` to only accept `dict` instances and `values_list.contains` to only accept `tuple` instances.

Reply:
@charettes As you might have seen, I posted the question to the mailing list. No response yet, which I intepret as a low interest.

I don't think the case I layed out above, wanting to check for presence in grouped QuerySet, is important. Furthermore, if we DO support that, we would make it hard to change the behaviour, while throwing an exception will keep that door open.

Therefore, I think it's wise to go with raising an exception for now. Error out, as you put it.

Reply:
@charettes Can you check if the new version is satisfactory?